app/src/main/java/com/github/campus_capture/bootcamp/map/MapScheduler.java

```diff
@@ -0,0 +1,259 @@
+    private boolean isZoneOwned;
+    private boolean isTakeover;
+    private boolean hasAttacked;
+    private Map<String, Section> zoneState;
+    private CountDownTimer buttonTimer;
+    public static boolean overrideTime = false;
+    public static Calendar time;
+
+    // Task to refresh the zone display at the top of the map
+    private final Runnable zoneRefreshTask = new Runnable() {
+        @Override
+        public void run() {
+            String label = upper.getString(R.string.current_zone_text);
+            LatLng position = upper.getCurrentPosition();
+            if(position == null)
+            {
+                label += "Unknown";
+                hideButtons();
+                scheduledTaskHandler.postDelayed(zoneRefreshTask, NO_POSITION_RETRY_DELAY);
+            }
+            else
+            {
+                Zone currentZone = upper.findCurrentZone(position);
+                if(currentZone == null)
+                {
+                    label += "None";
+                    isZoneOwned = false;
+                    hideButtons();
+                }
+                else
+                {
+                    String zoneName = currentZone.getName();
+                    label += zoneName;
+                    isZoneOwned = zoneState.get(zoneName) == User.getSection();
+                    showButtons();
+                }
+                scheduledTaskHandler.postDelayed(zoneRefreshTask, ZONE_REFRESH_RATE);
+            }
+            zoneText.setText(label);
+            Log.i("MapsFragment", "Refreshing current zone");
+        }
+    };
+
+    // Task to refresh the zone owners
+    private final Runnable refreshZoneState = new Runnable() {
+        @Override
+        public void run()
+        {
+            CompletableFuture<Map<String, Section>> val = CompletableFuture.supplyAsync(firebaseInterface::getCurrentZoneOwners);
+            try
+            {
+                zoneState = val.get();
+            }
+            catch(Exception e)
+            {
+                Log.e("MapScheduler", "Error ocurred when retrieving the zone owners");
+            }
+        }
+    };
+
+    // Task to open the attack buttons at the start of the takeover
+    private final Runnable openAttacksTask = new Runnable() {
+        @Override
+        public void run() {
+            isTakeover = true;
+            hasAttacked = false;
+            showButtons();
+            scheduledTaskHandler.postDelayed(closeAttacksTask, TAKEOVER_DURATION);
+        }
+    };
+
+    // Task to close the attack buttons once the takeover is over
+    private final Runnable closeAttacksTask = new Runnable() {
+        @Override
+        public void run() {
+            isTakeover = false;
+            showButtons();
+            scheduledTaskHandler.postDelayed(openAttacksTask, MILLIS_PER_HOUR - TAKEOVER_DURATION);
+            scheduledTaskHandler.postDelayed(refreshZoneState, OWNER_REFRESH_DELAY);
+        }
+    };
+
```

Comment:
Those are Runnables and  not functions cause they need to be used by an asynchron scheduler right ?

Reply:
Correct! There are some other ways to solve this but this was mainly done to alleviate the methods, and since the tasks do not change it makes more sense to store them locally.